src/descriptor/tr.rs

```diff
@@ -0,0 +1,279 @@
+pub struct Tr<Pk: MiniscriptKey> {
+    internal_key: Pk,
+    tree: Option<TapTree<Pk>>,
+}
+
+impl<Pk: MiniscriptKey> TapTree<Pk> {
+    pub fn to_string_no_checksum(&self) -> String {
+        match self {
+            TapTree::Tree(ref left, ref right) => {
+                format!("{{{},{}}}", *left, *right)
+            }
+            TapTree::Leaf(ref script) => format!("{}", *script),
+        }
+    }
+}
+
+impl<Pk: MiniscriptKey> fmt::Display for TapTree<Pk> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let desc = self.to_string_no_checksum();
+        write!(f, "{}", &desc)
+    }
+}
+
+impl<Pk: MiniscriptKey> Tr<Pk> {
+    fn taptree_height(tree: &TapTree<Pk>) -> usize {
+        match tree {
+            TapTree::Tree(ref left_tree, ref right_tree) => {
+                1 + max(
+                    Self::taptree_height(&**left_tree),
+                    Self::taptree_height(&**right_tree),
+                )
+            }
+            TapTree::Leaf(_) => 1,
+        }
+    }
+
+    pub fn new(internal_key: Pk, tree: Option<TapTree<Pk>>) -> Result<Self, Error> {
+        let nodes = match tree {
+            Some(ref t) => Self::taptree_height(t),
+            None => 0,
+        };
+
+        if nodes > TAPROOT_MAX_NODE_COUNT {
+            Ok(Self { internal_key, tree })
+        } else {
+            Err(Error::MaxRecursiveDepthExceeded)
+        }
+    }
+
+    fn to_string_no_checksum(&self) -> String {
+        let key = &self.internal_key;
+        match self.tree {
+            Some(ref s) => format!("tr({},{})", key, s),
+            None => format!("tr({})", key),
+        }
+    }
+
+    pub fn get_internal_key(&self) -> Result<Pk, Error> {
+        Ok(self.internal_key.clone())
+    }
+
+    pub fn get_taptree(&self) -> Result<TapTree<Pk>, Error> {
+        match &self.tree {
+            Some(t) => Ok(t.clone()),
+            None => Err(Error::Unexpected("TapTree empty".to_string())),
+        }
+    }
+}
+
+impl<Pk> Tr<Pk>
+where
+    Pk: MiniscriptKey + FromStr,
+    Pk::Hash: FromStr,
+    <Pk as FromStr>::Err: ToString,
+    <<Pk as MiniscriptKey>::Hash as FromStr>::Err: ToString,
+{
+    fn parse_miniscript(script: &str) -> Result<Miniscript<Pk, Segwitv0>, Error> {
+        let (ref script_tree, rest) = Tree::from_slice(script)?;
+        if rest.is_empty() {
+            Miniscript::<Pk, Segwitv0>::from_tree(script_tree)
+        } else {
+            return Err(Error::Unexpected(format!(
+                "error parsing miniscript from tapscript"
+            )));
+        }
+    }
+
+    // helper function for semantic parsing of script paths
+    pub fn tr_script_path(tree: &Tree) -> Result<TapTree<Pk>, Error> {
+        match tree {
+            Tree { name, args } if name.len() > 0 && args.len() == 0 => {
+                let script = name;
+                let script = Self::parse_miniscript(script)?;
+                let script = Arc::new(script);
```

Comment:
nit: slight code preference for inlining some operations. You don't really need the function `parse_miniscript`. You can call 
```
Miniscript::<Pk, Ctx>::from_str()
```
to get the miniscript directly. The function already does all the checks that you implement here. Also, the method `parse_method` does make sense itself inside an impl for `TapTree`